c++/src/capnp/compat/byte-stream.c++

```diff
@@ -756,59 +766,37 @@ public:
           });
         } else {
```

Comment:
The different behavior here being that if it's `MAX_BYTES_PER_WRITE`, it should write N times (up to `kjStream.limit`) before returning the stream?

Reply:
It wouldn't return the borrowed stream after the first write. But it would recurse to `write()` which would eventually release the borrowed stream once that limit is reached, yes.

Reply:
Oh boy... let me see here... :upside_down_face:

Reply:
Er, I just pushed something but now on re-reading `returnStream` it looks like it might not be doing what I expected (on my first read I thought it wasn't actually doing the return until `completed` bytes were written.

https://github.com/capnproto/capnproto/blob/29cc8acf67e28504d60f12d169507ece58528ce0/c%2B%2B/src/capnp/compat/byte-stream.c%2B%2B#L80-L89

Reply:
I should have said I updated this and think it's ready. I'm not in love with the `written` field on the `BorrowedStream` but I don't know where else it could live that'd be cleaner. My only other thought was to make `BorrowedStream` itself an `AsyncOutputStream` that would implement the writes so it could handle things, but I think that'd be moving a lot of stuff around for little gain (and probably has other issues I haven't thought of yet).

Could be missing something simpler, though. :)

Reply:
Heh, I just spent like 15 minutes being horribly confused by my own code here, but I think I finally figured it out.

So what's actually happening is, when `write()` first starts and calls `getShortestPath()`, that is what actually borrows the destination stream. Hence, you always need to return it at the end of the operation, and _before_ recursing. When you recurse into `write()` again, it's going to borrow the stream again, and it can't double-borrow.

I think your approach of adding `written` doesn't actually work then, because `kjStream` is actually a temporary so it's just getting thrown away when you recurse. But you don't need to accumulate. You should just always call `returnStream()` telling it exactly how much was written, before recursing.

But I guess that means `splitAndWrite()` can't actually directly recurse to `write()` because it doesn't know if it needs to call `returnStream()` first... But I think what you can do, actually, is undo the whole `splitAndWrite()` refactoring. All you really have to do is change the original code to use `kj::min(kjStream.limit, MAX_BYTES_PER_WRITE)` instead of `kjStream.limit` -- and then pass the actual amount written to `returnStream()`. If it's less that `kjStream.limit`, then the next `getShortestPath()` will borrow the same stream again and everything works fine.

Arguably it would be better to avoid returning and re-borrowing the same stream repeatedly, but for large writes I think it won't be a big deal.

Reply:
Sorry, my comment yesterday is totally wrong. I forgot that you're using `splitAndWrite()` in the `capnp::ByteStream::Client*` case, which didn't do any splitting before.

Reply:
Done. I definitely pushed my fixup (and can see as much in my shell history) but I'm failing to find it in the GitHub UI after my force push...

```diff
d73ebb75 - (origin/bhoerner/http-over-capnp-split-writes) fixup! Split very large byte-stream writes into smaller writes (4 minutes ago) <Brett Hoerner>
diff --git a/c++/src/capnp/compat/byte-stream.c++ b/c++/src/capnp/compat/byte-stream.c++
index ce77b408..c7b709dc 100644
--- a/c++/src/capnp/compat/byte-stream.c++
+++ b/c++/src/capnp/compat/byte-stream.c++
@@ -43,7 +43,6 @@ public:
     StreamServerBase& lender;
     kj::AsyncOutputStream& stream;
     uint64_t limit;
-    uint64_t written = 0;
   };
 
   typedef kj::OneOf<kj::Promise<void>, capnp::ByteStream::Client*, BorrowedStream> ShortestPath;
@@ -718,20 +717,16 @@ public:
         });
       }
       KJ_CASE_ONEOF(kjStream, StreamServerBase::BorrowedStream) {
-        auto limit = kj::min(kjStream.limit - kjStream.written, MAX_BYTES_PER_WRITE);
+        auto limit = kj::min(kjStream.limit, MAX_BYTES_PER_WRITE);
         if (size <= limit) {
           auto promise = kjStream.stream.write(buffer, size);
           return promise.then([kjStream,size]() mutable {
-            kjStream.written += size;
-            kjStream.lender.returnStream(kjStream.written);
+            kjStream.lender.returnStream(size);
           });
         } else {
           auto promise = kjStream.stream.write(buffer, limit);
           return promise.then([this,kjStream,buffer,size,limit]() mutable {
-            kjStream.written += limit;
-            if (kjStream.written == kjStream.limit) {
-              kjStream.lender.returnStream(kjStream.written);
-            }
+            kjStream.lender.returnStream(limit);
             return write(reinterpret_cast<const byte*>(buffer) + limit,
                          size - limit);
           });
@@ -770,18 +765,14 @@ public:
         if (size <= limit) {
           auto promise = kjStream.stream.write(pieces);
           return promise.then([kjStream,size]() mutable {
-            kjStream.written += size;
-            kjStream.lender.returnStream(kjStream.written);
+            kjStream.lender.returnStream(size);
           });
         } else {
           // ughhhhhhhhhh, we need to split the pieces.
           return splitAndWrite(pieces, kjStream.limit,
               [kjStream,limit](kj::ArrayPtr<const kj::ArrayPtr<const byte>> pieces) mutable {
             return kjStream.stream.write(pieces).then([kjStream,limit]() mutable {
-              kjStream.written += limit;
-              if (kjStream.written == kjStream.limit) {
-                kjStream.lender.returnStream(kjStream.written);
-              }
+              kjStream.lender.returnStream(limit);
             });
           });
         }
@@ -1001,8 +992,8 @@ private:
       // Oh thank god, the split is between two pieces.
       auto rest = pieces.slice(splitPiece, pieces.size());
       return writeFirstPieces(pieces.slice(0, splitPiece))
-          .then([this,rest,limit,writeFirstPieces]() mutable {
-        return splitAndWrite(rest, limit, writeFirstPieces);
+          .then([this,rest]() mutable {
+        return write(rest);
       });
     } else {
       // FUUUUUUUU---- we need to split one of the pieces in two.
```